inspirehep/utils/date.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2016 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+"""Helpers for handling dates."""
+
+import six
+
+from invenio_utils.date import strftime, strptime   # pre-1900 friendly parser
+
+DATE_FORMATS_YEAR = ["%Y", "%y"]
+DATE_FORMATS_MONTH = [
+    "%Y-%m", "%Y %b", "%b %Y", "%Y %B", "%B %Y",
+    "%y-%m", "%y %b", "%b %y", "%y %B", "%B %y",
+]
+DATE_FORMATS_FULL = [
+    "%Y-%m-%d", "%d %m %Y", "%x", "%d %b %Y",
+    "%d %B %Y", "%d %b %y", "%d %B %y",
+]
+
+
+def create_earliest_date(dates):
+    """Add '01' as a value of missing month and/or day values."""
+    if not dates:
+        return None
+    if not isinstance(dates, (list, set)):
+        dates = [dates]
+    valid_dates = []
+    for date in dates:
+        # Add 99 to the end of partial date to make sure we get the REAL
+        # earliest date
+        valid_date = create_valid_date(date,
+                                       date_format_month="%Y-%m-99",
+                                       date_format_year="%Y-99-99")
+        if valid_date:
+            valid_dates.append(valid_date)
+    if valid_dates:
+        date = min(valid_dates)
+        return date.replace('-99', '')
+
+
+def create_valid_date(date, date_format_full="%Y-%m-%d",
+                      date_format_month="%Y-%m", date_format_year="%Y"):
+    """Iterate over possible formats and return a valid date if found."""
+    valid_date = None
+    date = six.text_type(date)
+    for format in DATE_FORMATS_FULL:
+        try:
+            valid_date = strftime(date_format_full, (strptime(date, format)))
+            break
+        except ValueError:
+            pass
+    else:
+        for format in DATE_FORMATS_MONTH:
+            try:
+                if date.count('-') > 1:
+                    date = "-".join(date.split('-')[:2])
```

Comment:
Yeah, so it depends on the strategy we apply. I was discussing with @jmartinm in that we should really start validating the records with the schema and report any inconsistencies even before any of this "correcting" code is run. So I was thinking of adding a "non-breaking" schema validation to the migrator so that bad metadata will be logged but the record is still migrated and finally indexed. E.g. by adding some new columns to the prod migration model: `valid=True/False` and text with errors.

This code is now executed upon indexing with the goal of creating an indexable document. So one could in many cases save some part of a date that is malformed (say `2012-03-022`) would be returned as `2012-03`, until the date is fixed by a human after receiving the report from the schema validation.

PS: There may perhaps even exist some "correcting" code in the dojson that ideally should be reporting the bad metadata as well.